src/analysis/constant_speed.lean

```diff
@@ -0,0 +1,224 @@
+/-
+Copyright (c) 2023 Rémi Bottinelli. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Rémi Bottinelli
+-/
+import measure_theory.measure.lebesgue
+import analysis.calculus.monotone
+import data.set.function
+import analysis.bounded_variation
+import tactic.swap_var
+/-!
+# Constant speed
+
+This file defines the notion of constant (and unit) speed for a function `f : ℝ → E` with
+pseudo-emetric structure on `E` with respect to a set `s : set ℝ` and "speed" `l : ℝ≥0`, and shows
+that if `f` has locally bounded variation on `s`, it can be obtained (up to distance zero, on `s`),
+as a composite `φ ∘ (variation_on_from_to f s a)`, where `φ` has unit speed and `a ∈ s`.
+
+## Main definitions
+
+* `has_constant_speed_on_with f s l`, stating that the speed of `f` on `s` is `l`.
+* `has_unit_speed_on f s`, stating that the speed of `f` on `s` is `1`.
+* `natural_parameterization f s a : ℝ → E`, the unit speed reparameterization of `f` on `s` relative
+  to `a`.
+
+## Main statements
+
+* `unique_unit_speed_on_Icc_zero` proves that if `f` and `f ∘ φ` are both naturally
+  parameterized on closed intervals starting at `0`, then `φ` must be the identity on
+  those intervals.
+* `natural_parameterization_edist_zero` proves that if `f` has locally bounded variation, then
+  precomposing `natural_parameterization f s a` with `variation_on_from_to f s a` yields a function
+  at distance zero from `f` on `s`.
+* `natural_parameterization_has_unit_speed` proves that if `f` has locally bounded
+  variation, then `natural_parameterization f s a` has unit speed on `s`.
+
+## Tags
+
+arc-length, parameterization
+-/
+
+open_locale big_operators nnreal ennreal
+open set measure_theory classical
+
+variables {α : Type*} [linear_order α] {E : Type*} [pseudo_emetric_space E]
+variables (f : ℝ → E) (s : set ℝ) (l : ℝ≥0)
+
+/--
+`f` has constant speed `l` on `s` if the variation of `f` on `s ∩ Icc x y` is equal to
+`l * (y - x)` for any `x y` in `s`.
+-/
+def has_constant_speed_on_with :=
+∀ ⦃x⦄ (hx : x ∈ s) ⦃y⦄ (hy : y ∈ s), evariation_on f (s ∩ Icc x y) = ennreal.of_real (l * (y - x))
+
+variables {f} {s} {l}
+
+lemma has_constant_speed_on_with.has_locally_bounded_variation_on
+  (h : has_constant_speed_on_with f s l) : has_locally_bounded_variation_on f s := λ x y hx hy,
+by simp only [has_bounded_variation_on, h hx hy, ne.def, ennreal.of_real_ne_top, not_false_iff]
+
```

Comment:
can you add a lemma `has_constant_speed_on_with.mono`, saying that if `f` has constant speed on some set `s`, and `t` is contained in `s`, then `f` has also constant speed on `t`? Same thing for `has_unique_speed`. Also another lemma saying that if you have constant speed `l` on two sets `s` and `t`, and a point `x` is the lub of `s` and the glb of `t`, then `f` has constant speed `l` on their union? And the obvious consequence for two intervals? And the same thing for unit speed? I know this is cumbersome, but it's the point of API building to provide all the basic lemmas like that.

Reply:
The mono version can be proven like:
```
lemma has_constant_speed_on.mono {t : set ℝ} (ts : t ⊆ s)
  (hts : ∀ ⦃x⦄ (hx : x ∈ t) ⦃y⦄ (hy : y ∈ t), t ∩ (Icc x y) = s ∩ (Icc x y))
  (hfs : has_constant_speed_on_with f s l) : has_constant_speed_on_with f t l :=
λ x hx y hy, by rw [←hfs (ts hx) (ts hy), ←hts hx hy]
```
which I'm not sure is worth it? I'll add it anyway and let you judge.

Reply:
I added the other lemmas, with similar signatures to what you had for `evariation_on` (so not lub/glb but just greatest/least).
Actually, the lub/glb version is probably wrong, letting `s = Ico 0 1` and `t = Ioc 1 2` and a jump at `1`, makes things wrong.

Reply:
Yes, I meant greatest and least, sorry.